crates/noirc_evaluator/src/ssa_refactor/opt/defunctionalize.rs

```diff
@@ -0,0 +1,364 @@
+///     }
+/// }
+/// ```
+/// Apply functions generally take the function to apply as their first parameter. This is a Field value
+/// obtained by converting the FunctionId into a Field. The remaining parameters of apply are the
+/// arguments to forward to this function when calling it internally.
+#[derive(Debug, Clone, Copy)]
+struct ApplyFunction {
+    id: FunctionId,
+    dispatches_to_multiple_functions: bool,
+}
+
+/// Performs defunctionalization on all functions
+/// This is done by changing all functions as value to be a number (FieldElement)
+/// And creating apply functions that dispatch to the correct target by runtime comparisons with constants
+#[derive(Debug, Clone)]
+struct DefunctionalizationContext {
+    fn_to_runtime: HashMap<FunctionId, RuntimeType>,
+    variants: HashMap<FunctionSignature, Vec<FunctionId>>,
+    apply_functions: HashMap<FunctionSignature, ApplyFunction>,
+}
+
+impl Ssa {
+    pub(crate) fn defunctionalize(mut self) -> Ssa {
+        // Find all functions used as value that share the same signature
+        let variants = find_variants(&self);
+
+        let apply_functions = create_apply_functions(&mut self, &variants);
+        let fn_to_runtime =
+            self.functions.iter().map(|(func_id, func)| (*func_id, func.runtime())).collect();
+
+        let context = DefunctionalizationContext { fn_to_runtime, variants, apply_functions };
+
+        context.defunctionalize_all(&mut self);
+        self
+    }
+}
+
+impl DefunctionalizationContext {
+    /// Defunctionalize all functions in the Ssa
+    fn defunctionalize_all(mut self, ssa: &mut Ssa) {
+        for function in ssa.functions.values_mut() {
+            self.defunctionalize(function);
+        }
+    }
+
+    /// Defunctionalize a single function
+    fn defunctionalize(&mut self, func: &mut Function) {
+        let mut target_function_ids = HashSet::new();
+
+        for block_id in func.reachable_blocks() {
+            let block = &func.dfg[block_id];
+            let instructions = block.instructions().to_vec();
+
+            for instruction_id in instructions {
+                let instruction = func.dfg[instruction_id].clone();
+                let mut replacement_instruction = None;
+                // Operate on call instructions
+                let (target_func_id, arguments) = match instruction {
+                    Instruction::Call { func: target_func_id, arguments } => {
+                        (target_func_id, arguments)
+                    }
+                    _ => continue,
+                };
+
+                match func.dfg[target_func_id] {
+                    // If the target is a function used as value
+                    Value::Param { .. } | Value::Instruction { .. } => {
+                        // Collect the argument types
+                        let argument_types: Vec<Type> =
+                            arguments.iter().map(|arg| func.dfg.type_of_value(*arg)).collect();
```

Comment:
You don't need to get rid of the vecmap, you can pass arguments by reference:
```suggestion
                        let argument_types = vecmap(&arguments, |arg| func.dfg.type_of_value(*arg));
```